go/vt/vtgate/engine/concatenate.go

```diff
@@ -0,0 +1,209 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"sort"
+	"strings"
+	"sync"
+
+	"vitess.io/vitess/go/mysql"
+
+	"vitess.io/vitess/go/sqltypes"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+// Concatenate Primitive is used to concatenate results from multiple sources.
+var _ Primitive = (*Concatenate)(nil)
+
+//Concatenate specified the parameter for concatenate primitive
+type Concatenate struct {
+	Sources []Primitive
+}
+
+//RouteType returns a description of the query routing type used by the primitive
+func (c *Concatenate) RouteType() string {
+	return "Concatenate"
+}
+
+// GetKeyspaceName specifies the Keyspace that this primitive routes to
+func (c *Concatenate) GetKeyspaceName() string {
+	ksMap := map[string]interface{}{}
+	for _, source := range c.Sources {
+		ksMap[source.GetKeyspaceName()] = nil
+	}
+	var ksArr []string
+	for ks := range ksMap {
+		ksArr = append(ksArr, ks)
+	}
+	sort.Strings(ksArr)
+	return strings.Join(ksArr, "_")
+}
+
+// GetTableName specifies the table that this primitive routes to.
+func (c *Concatenate) GetTableName() string {
+	var tabArr []string
+	for _, source := range c.Sources {
+		tabArr = append(tabArr, source.GetTableName())
+	}
+	return strings.Join(tabArr, "_")
+}
+
+// Execute performs a non-streaming exec.
+func (c *Concatenate) Execute(vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
+	result := &sqltypes.Result{}
+	var wg sync.WaitGroup
+	qrs := make([]*sqltypes.Result, len(c.Sources))
+	errs := make([]error, len(c.Sources))
+	for i, source := range c.Sources {
+		wg.Add(1)
+		go func(i int, source Primitive) {
+			defer wg.Done()
+			qrs[i], errs[i] = source.Execute(vcursor, bindVars, wantfields)
+		}(i, source)
+	}
+	wg.Wait()
+	for i := 0; i < len(c.Sources); i++ {
+		if errs[i] != nil {
+			return nil, vterrors.Wrap(errs[i], "Concatenate.Execute")
+		}
+		qr := qrs[i]
+		if result.Fields == nil {
+			result.Fields = qr.Fields
+		}
+		err := compareFields(result.Fields, qr.Fields)
+		if err != nil {
+			return nil, err
+		}
+		if len(qr.Rows) > 0 {
+			result.Rows = append(result.Rows, qr.Rows...)
+			if len(result.Rows[0]) != len(qr.Rows[0]) {
+				return nil, mysql.NewSQLError(mysql.ERWrongNumberOfColumnsInSelect, "21000", "The used SELECT statements have a different number of columns")
+			}
+			result.RowsAffected += qr.RowsAffected
+		}
+	}
+	return result, nil
+}
+
+// StreamExecute performs a streaming exec.
+func (c *Concatenate) StreamExecute(vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, callback func(*sqltypes.Result) error) error {
+	var seenFields []*querypb.Field
+	var fieldset sync.WaitGroup
+	fieldsSent := false
+
+	g := vcursor.ErrorGroupCancellableContext()
```

Comment:
I think we need the engine primitives to use this new cancelable context. Contexts are like onions - there are layers to them. We are merely adding a cancelable layer to the context the caller provided us with. if you want, we could peel that layer before leaving this method, to restore the original context.

Reply:
Oh. I didn't realize that you were passing vcursor down to the callees. So, it's ok to leave it as is for now.

Reply:
The context created here needs to be sent down to calls from here one. So that once the context is cancelled. The lower layers can also stop processing.
The issue here is the lower layer can receive this new context via vcursor as of now. So, the current way was to replace it.
Once we start passing context separately this would not be an issue.

@systay The current suggestion does not hold. Because if we replace the vcursor context and there is parent branch which has vcursor usage. It will already be updated and using the new context. So peeling off before returning may not guarantee things are in order.